### ISSUE TITLE
List and recognize ClassicPress contributors

### ISSUE THREAD

**Issue by nylen** [ClassicPress/ClassicPress#284]
We need to come up with a better way to list and recognize ClassicPress contributors.

Previously we were using OpenCollective to show a list of all code contributors, but as pointed out in #271, this shows past WordPress contributors together with active ClassicPress contributors, which is not ideal.

Ideally our contributions list should include people who have participated in discussions, helped with marketing efforts, donated, etc.

GitHub is a good data source for a lot of these items.  Commits, issue and PR body texts, discussion and review comments, etc can be fetched using their API.

Once we build something to collect this data, we can also use it to assign badges on [our forums](https://forums.classicpress.net/).

See also:

- https://opencollective.com/classicpress/contributors.svg?width=890&button=false (the previous list of contributors)
- https://github.com/opencollective/opencollective-images/blob/2ab6fdf92e8496465bce507fe22fe2fd7b574895/src/server/controllers/collectives.js
- https://github.com/ClassicPress/ClassicPress/compare/LAST_WP_COMMIT...develop (a list of code contributions to ClassicPress)
- https://github.com/discourse/discourse-github/blob/50c42d8a0b39106bce72935eb8c216a5c522b50d/app/lib/github_badges.rb (a Discourse plugin to add badges based on the history of a `git` repository)

**Comment by Mte90** [comment]
I know that in GitHub is possible to reset the list of contributors from a specific date.
Contacting the support they can do that so we will have a reset point but for the future when we merge commits this will be a little problem.

**Comment by nylen** [comment]
We'll need to build something to recognize "Props" lines in commit messages.  For example, from 36152b6:

> Props @Pross.

**Comment by pattonwebz** [comment]
History squashing to clear out props to past WP contributors may be possible on the repos history but I would encourage you NOT to do that. 

While those people are not contributors to classicpress right now they still were in the past (albeit via CP forking their work) and as such they still deserve props and recognition. Also I believe morally it is much preferable to retain original credits (and indeed a requirement of the GPL freedoms that we both are granted and must offer).

**Comment by nylen** [comment]
>History squashing to clear out props to past WP contributors may be possible on the repos history but I would encourage you NOT to do that.

Agreed, we are definitely not going to rewrite the commit history.

I think the discussion above refers to asking GitHub to only count contributions in the [graphs](https://github.com/ClassicPress/ClassicPress/pulse) after our fork point, which they can do, but I'm not convinced this is necessary or valuable either.

**Comment by nylen** [comment]
As a specific example of something we need to be able to do here, ideally in a (semi-)automated fashion:

#419 was written by @bahiirwa , and he used code (patches) by https://profiles.wordpress.org/redsweater/ (Daniel Jalkut (Red Sweater)) in the PR.  These patches are originally from https://core.trac.wordpress.org/ticket/45322.

We didn't think to record Daniel as the original author of the patch in the commit history, but we should definitely make note of this somehow.

I think a good solution would be a script that generates/updates a contributors list based on our commit history, and maybe we'd just manually edit the list for special situations like this.

cc @DanielJ

**Comment by elisabettac77** [comment]
What I think we can do is fetch all the raw data mapping user contribution statistics from all the sources (forums, Zulip, GitHub, Crowdin, OpenCollective, etc) via their own API (I checked and they all support that) and then create a system that awards points for each specific action. The system will update real time, if an user stops to contribute their points will lower and thus the ranking will too. As soon as they start contributing again their points will go up again. the front-end could show ranking based on points, and allow filtering of contributors that are also "service providers" within the community and show on a detail page what each user has contributed to.
This system would be similar to what Drupal does, and was inspired by it.
It is not only a way to give props and recognition, but also to showcase the ones who can offer CP related services so that users in need of a professional can have a place to refer to.
I am working on this but the conversion chart (actions==>points) needs to be established by CP community as a whole, I can't just arbitrarily decide on my own how many points a merged PR or a solved issue on forums is valued. I can give a list of actions that I can retrieve fro the various platforms that makes sense to be tracked then we could discuss their value and implement it so that my system converts actions into points.
To be noted that the only value of such points would be recognition within the community, and that only consistent contributions will allow people to show on the list (this means there's no real way to circumvent the system given the fact it takes data via API real time) and if that is not enough CP Initiative can supervise the list to guarantee that is fairly managed.

**Comment by ginsterbusch** [comment]
Thats a nice idea, but there are a few concerns:

1. for one, data privacy, as this is mass tracking and surveilance.. 

2. and the other: so, as a person who engages mostly outside of this exclusive circle, I would be basically punished? Because there are a lot of individuals who dont (massivelly) interact within the these "hubs", and still "contribute". So some sort of forced interaction? Else you're not deemed a good person for this "community"?

I guess mass surveilance and tracking and forced interaction is not good for true community building. Its the opposite. Also see "Überwachungsstaat", "Totalitarian Regimes", "Marxism-Leninism", "Stasi" and all the other nasty examples of history :)

It certainly would drive me away.

**Comment by elisabettac77** [comment]
Thanks for the input @ginsterbusch - I really appreciate your insight on this.

1) privacy concern: It is indeed data tracking and statistic reporting.

Basically I extract the data that the platforms already report about users (and they already gave consent for this when using them) and show them in ranking form (we can decide to only show the ranking, not the actions to be more GDPR but for transparency we will need to make the data publicly available upon request all the same so we could just make it available and transparent). We also might need to be clear about that procedure with a wording on our privacy pages "Be aware that by interacting with CP ecosystem (list of CP platforms here) you accept that your contribution is tracked and reported with the goal to give credit to every contributor on our platform xxx".
That might need more input by @KTS915 since it touches the legal realm.
Privacy is a valid concern of mine too, however if Drupal was able to make this possible it means there is a way that is legal.

2) The goal of such a thing is not to quantify who contributes more or less. Nor is to punish people who interact less. (I am squeezing my brain on how to track social media interactions and other actions performed outside CP ecosystem of platforms that are valuable for CP in order to list also those people... if that makes sense!) - the goal of this is to collect in one place a very comprehensive list of ALL people who contribute to CP, no matter how they do it or what platform they use to do so.

To be more clear, we can track and report data in a way that is privacy compliant. and we can even have a simple list of ALL the contributors that gravitate around CP Ecosystem without ranking them. We can also just show that they contributed without publicly disclosing what they contributed or how much. We could just have the list and a badge to specify which ones are contributors that offer CP related services. The real issue for me at present is finding a way to capture contributions from people who interact less on the official platforms like GitHub, Forum, Zulip, Crowdin or Directory (that are the ones that are easily solved because they already have statistic data available that can be retrieved via API).

Do we even want to track who contributes by advocating for CP on social media for example? It would fall under Marketing Efforts and it would make sense to track it someway? Or do we only track the easy ones? that would possibly leave some people out of the recognition they deserve.

**Comment by ginsterbusch** [comment]
In legal terms, if you modify TOS or privacy regulations AFTER the user has agreed to them, you have to tell them.
Also, this specific kind of tracking requires options to disable it, because it is not an essential part of the platform.

"privacy compliant", esp. by European laws, eg. the GDPR, also includes ensuring not to record data more than necessary, and also, 

I strongly suggest looking into topics like Consent Management.

Except if you want to exclude european users (Including me, and possibly yourself) from using CP as a community platform.

And no, I've not consented to share my data with any external parties **outside** of Github. Just because something is publicy accessible doesnt mean you're also allowed to use it freely :)

cu, w0lf.

**Comment by elisabettac77** [comment]
@pattonwebz I do like the idea of tracking contributors, but I have an issue with tracking contributors of WP patches. If I understand correctly the aim of this Issue is to track only the people who directly contribute to CP, since the WP contributors we "inherit" from backports and patches are however listed in changelog as such.

Agreed on the fact we do not need the ranking.

About how to track contributions that happen outside CP Platforms, that is the big issue I am trying to understand if it is worth solving. As you say it might be an important thing to do so.

**Comment by elisabettac77** [comment]
@ginsterbusch I do agree with you that IF we decide to track contributions (and in some way we might track them since all FOSS projects do) it is mandatory that we change our policies first and give an appropriate communication in advance about that.
It is obviously clear that consent to be tracked as contributors is optional (that means if one does not grant it it does not appear).
Your concern about privacy is very valid.
As I said however all FOSS projects do track their contributors in a way or another. It is something that in the field seems to be "expected" in a way. Do they all do that by violating the GDPR and World Wide Privacy Regulations? The need to change our policies and to give a decent communication in advance about that to contributors and users should not hinder our ability to make this thing happen IMHO.

**Comment by ginsterbusch** [comment]
Just because something is "expected" doesnt make it better, no? And also not more legal (and even if something is deemed "legal", doesnt make it automatically a "good thing") :)

Also, what does "all FOSS projects" mean? Thats just a generalization and assumption of yours.
There is a multitude of legal and privacy frameworks, allowances, regulation, consent management going on, that one cannot just stuff into "all FOSS projects do it this way".

Hence why: Look into consent management.

Just my autistic 2 cents to this topic.

**Comment by KTS915** [comment]
Before we get into the legalities of tracking "contributions" in all their forms, I'd want us to be clear about what it is we'd be trying to do and why.

As a matter of principle, I don't like the idea of tracking contributions (and, therefore, contributors) for two reasons:

1. It seems to me to be attempting to solve a problem that ClassicPress doesn't have.
2. There's no meaningful metric that can capture the quality or value of a contribution as opposed to the number of contributions.

So I think this suggestion would create problems where none currently exist. I'm also not sold on using Drupal as a model of governance. That also effectively has a so-called BDFL (though they never turn out to be benevolent), and version 8 of Drupal was met by its user-base in much the same way that Gutenberg has been received by WP users. (You can see that from the Drupal releases page at https://www.drupal.org/project/drupal/releases, where you'll find that it has essentially fractured into two: version 7 for those not wanting to go to version 8 vs version 10 (or 11 for early adopters) for those who took the plunge.)

I'm all for making it easier to find a list of everyone who has contributed to each release, so that it's easier to list them when a release is made. What that requires is some code to extract names from all those who have submitted or commented on PRs. But I don't think tracking anyone is either necessary or helpful.

**Comment by elisabettac77** [comment]
>Just because something is "expected" doesnt make it better, no? And also not more legal (and even if something is >deemed "legal", doesnt make it automatically a "good thing") :)

On that I can agree to a certain extent. However the fact it is we are not discussing the good and the bad. We are discussing how do we set up and implement REALISTICALLY a way to give credit/thank/make visible the people (ideally all of them and that is where the thing crashes into the realistic side) that make CP.

>Also, what does "all FOSS projects" mean? Thats just a generalization and assumption of yours.
What I mean here is that pretty much all projects maintain somewhere a list of people who contributed. Clearly they manage consent as you are advising (and that is why I say that by managing consent we could find a way to do this in a way that gives recognition AND is also legal)

On if it is good or bad to give recognition, IMHO this is an opinion. My opinion is that having a list of contributors to thank for CP might be good. My vote counts for one however. :)

>There is a multitude of legal and privacy frameworks, allowances, regulation, consent management going on, that one >cannot just stuff into "all FOSS projects do it this way".

VERY TRUE. That is why we need to find a way to do the consent management the right way and I tagged Tim on this also.

>Hence why: Look into consent management.

See above :)

On the value of two cents, literally it could seem nothing. Think about what would CP be without people having the will to throw them in however.

To bring my feet on the ground again, before even starting to do something I have to set up a way to AT LEAST extract the names of people who contribute. And if possible not only from the platforms that make it easy for us with APIs, but find a way to do so from other places too (and do that in a way that consent can be tracked/managed by users in a correct and legal way). I think that once this is done, we will REALLY need to set up consent management. and then think about implementing the thing where it is more appropriate to do so.

**Comment by elisabettac77** [comment]
@KTS915 Thanks for your input.

I am favorable to contribution tracking for these reasons:

- It helps show the project is alive and thriving
- It encourages people to contribute too
- It markets the software basically (being an action of brand -awareness that we could undertake)
- It helps people in feeling recognized and motivated in the long run, to continue contributing
- It shows community cares about community
- It might serve also to show the project's growth
- It might be something we might need to do in the future when community is larger to track who worked/works on what

I think only tracking the few that take care of releases means leaving behind all the people who everyday develop software for CP, volunteer in localizing or help with brand awareness, donate to the project or contribute in their own way to CP. CP is not just the people who make the releases (and credits to them are already automated on the welcome page when doing an upgrade and in changelog) - I hope we can at least agree on this.

About Drupal governance, I agree it's not an ideal solution for CP. I was not saying we should adopt that as a whole. Just the idea that we could have a way to showcase all contributors that make CP, and eventually highlight who among them is also a Service Provider.

About Drupal fiasco between v7 and v8, basically it was caused by same reasons WP fiasco was. Drupal is not in touch with its users and community at large, and this caused them to break promises to said community and user base. This brings me to my last strong point:

Listing contributors might also be a way to maintain contact with the community by showing that we value every input albeit small.

Obviously I am one. And as such I vote for one. So I ask again:

Is there any real interest in solving this Issue? Is it just a nice to have we are discussing that will drag for some other time? It's been created end 2018 and not solved yet. Can we decide if we go ahead with it (and in that case I am available to help) or we close it?

**Comment by KTS915** [comment]
I agree that _listing_ contributors is important. I disagree about the need to _track_ them.

I don't see any reason to close this Issue. We need to find a better way than the manual method we're using currently. Until we do, this remains a task to be addressed. If anyone could come up with a script for keeping a record of who proposes and comments on PRs, that would be great.

**Comment by xxsimoxx** [comment]
In the last release @mattyrob used [this plugin](https://github.com/xxsimoxx/cp-contributors) to list contributors.
It takes commits between two tags and extract authors and coauthors.
Also `WP:Props` are listed.
If we want we can also add `CP:Props` line to pull requests (this is a manual work) and it's easy to update the plugin to get also those.

**Comment by KTS915** [comment]
> If we want we can also add CP:Props line to pull requests (this is a manual work) and it's easy to update the plugin to get also those.

I'd say please do that!

**Comment by KTS915** [comment]
Excellent!

**Comment by mattyrob** [comment]
Propose we close this now we have and use the CP Contributors plugin.
